src/algebraic_geometry/open_immersion.lean

```diff
@@ -1062,4 +1062,156 @@ end pullback
 
 end LocallyRingedSpace.is_open_immersion
 
+lemma is_open_immersion.open_range {X Y : Scheme} (f : X ⟶ Y) [H : is_open_immersion f] :
+  is_open (set.range f.1.base) := H.base_open.open_range
+
+section open_cover
+
+namespace Scheme
+
+/-- An open cover of `X` consists of a family of open immersions into `X`,
+and for each `x : X` an open immersion (indexed by `f x`) that covers `x`.
+
+This is merely a coverage in the Zariski pretopology, and it would be optimal
+if we could reuse the existing API about pretopologies, However, the definitions of sieves and
+grothendieck topologies uses `Prop`s, so that the actual open sets and immersions are hard to
+obtain. Also, since such a coverage in the pretopology usually contains a proper class of
+immersions, it is quite hard to glue them, reason about finite covers, etc.
+-/
+-- TODO: provide API to and from a preseive.
+structure open_cover (X : Scheme.{u}) :=
+(J : Type u)
+(obj : Π (j : J), Scheme)
+(map : Π (j : J), obj j ⟶ X)
```

Comment:
Two thoughts that pop into my head as alternative design decisions:

* Make `J` a parameter instead of a structure field. (And if you want to index by `fin n`, maybe decouple the universe?)
* Alternative: bundle everything together as a function into the over category of `X` (i.e. Schemes over `X`).

Did you consider these? If so, why did you choose for the current design?

Reply:
I don't think we would need to require the open cover to be indexed by a specific Type. Most of the times, only the cardinality matters, and in the current design where the universe is fixed, the only useful distinction is whether `J` is finite or not, which could still be easily imposed via `[fintype U.J]`.
On the other hand, a lot of the times the construction of `J` would be quite ugly.
For example, this `t` is the `J` for a finite subcover of an open cover.
```lean
have := @@compact_space.elim_nhds_subcover _ H
    (λ (x : X.carrier), set.range ((𝒰.map (𝒰.f x)).1.base))
    (λ x, (is_open_immersion.open_range (𝒰.map (𝒰.f x))).mem_nhds (𝒰.covers x)),
  let t := this.some,
 ```
 I don't think it would be easy to write all this in the result type.

Reply:
On the other hand, the universe restriction was imposed so that this could connect to the gluing API easily.
I suppose we could generalize it and only impose the restriction when nessecary?

Reply:
I suppose only the `obj` and `map` field could be bundled together?
I don't think the over category API would be useful here. On the down side, `𝒰.obj` becomes `𝒰.map.left` which is probably not ideal?